File: practice/src/main/java/com/decade/practice/adapter/web/rest/ChatController.java
package com.decade.practice.adapter.web.rest;

import com.decade.practice.application.usecases.ChatService;
import com.decade.practice.application.usecases.DeliveryService;
import com.decade.practice.domain.ChatSnapshot;
import com.decade.practice.domain.embeddables.ChatIdentifier;
import com.decade.practice.domain.embeddables.Preference;
import com.decade.practice.domain.entities.Chat;
import com.decade.practice.domain.entities.PreferenceEvent;
import com.decade.practice.domain.entities.Theme;
import com.decade.practice.domain.entities.User;
import com.decade.practice.domain.repositories.ChatRepository;
import com.decade.practice.domain.repositories.ThemeRepository;
import com.decade.practice.domain.repositories.UserRepository;
import com.decade.practice.utils.WebCacheUtils;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.security.Principal;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/chats")
public class ChatController {

        private final UserRepository userRepository;
        private final ChatRepository chatRepository;
        private final ChatService chatService;
        private final ThemeRepository themeRepository;
        private final DeliveryService deliveryService;

        public ChatController(
                UserRepository userRepository,
                ChatRepository chatRepository,
                ChatService chatService,
                ThemeRepository themeRepository,
                DeliveryService deliveryService
        ) {
                this.userRepository = userRepository;
                this.chatRepository = chatRepository;
                this.chatService = chatService;
                this.themeRepository = themeRepository;
                this.deliveryService = deliveryService;
        }

        @GetMapping("/{identifier}")
        public ResponseEntity<ChatSnapshot> get(
                Principal principal,
                @PathVariable ChatIdentifier identifier,
                @RequestParam(required = false) Integer atVersion
        ) {
                User me = userRepository.findByUsername(principal.getName());
                Chat chat = chatService.getOrCreateChat(identifier);
                ChatSnapshot snapshot = chatService.getSnapshot(
                        chat,
                        me,
                        atVersion != null ? atVersion : me.getSyncContext().getEventVersion()
                );

                return ResponseEntity.ok()
                        .cacheControl(WebCacheUtils.ONE_MONTHS)
                        .header("Vary", "Cookie", "Authorization")
                        .body(snapshot);
        }

        @GetMapping("/partners/{partnerId}")
        public ResponseEntity<ChatSnapshot> get(
                Principal principal,
                @PathVariable UUID partnerId,
                @RequestParam(required = false) Integer atVersion
        ) {
                User me = userRepository.findByUsername(principal.getName());
                ChatIdentifier identifier = ChatIdentifier.from(partnerId, me.getId());
                Chat chat = chatService.getOrCreateChat(identifier);
                ChatSnapshot snapshot = chatService.getSnapshot(
                        chat,
                        me,
                        atVersion != null ? atVersion : me.getSyncContext().getEventVersion()
                );

                return ResponseEntity.ok()
                        .cacheControl(WebCacheUtils.ONE_MONTHS)
                        .header("Vary", "Cookie", "Authorization")
                        .body(snapshot);
        }

        @PatchMapping("/{identifier}/preference")
        public Preference updatePreferences(
                Principal principal,
                @PathVariable ChatIdentifier identifier,
                @RequestBody Preference preference) {

                User me = userRepository.findByUsername(principal.getName());

                Theme theme = preference.getTheme();
                if (theme != null) {
                        preference.setTheme(themeRepository.findById(theme.getId()).orElseThrow());
                }
                PreferenceEvent event = new PreferenceEvent();
                event.setChatIdentifier(identifier);
                event.setSender(me);
                event.setPreference(preference);
                deliveryService.createAndSend(me, event);

                return preference;
        }

        @GetMapping("/themes")
        public ResponseEntity<List<Theme>> getThemes() {
                return ResponseEntity.ok().cacheControl(
                                WebCacheUtils.ONE_MONTHS.cachePublic())
                        .body(themeRepository.findAll());
        }

        @GetMapping
        public ResponseEntity<List<ChatSnapshot>> list(
                Principal principal,
                @RequestParam(required = false) ChatIdentifier startAt,
                @RequestParam int atVersion
        ) {
                Chat chat = startAt == null ? null : chatRepository.findById(startAt).get();
                User owner = userRepository.findByUsername(principal.getName());

                List<Chat> chatList = chatService.listChat(owner, atVersion, chat);
                List<ChatSnapshot> snapshotList = new ArrayList<>();
                for (Chat c : chatList) {
                        snapshotList.add(chatService.getSnapshot(c, owner, atVersion));
                }
                return ResponseEntity.ok()
                        .cacheControl(WebCacheUtils.ONE_MONTHS)
                        .header("Vary", "Cookie", "Authorization")
                        .body(snapshotList);
        }
}

File: practice/src/main/java/com/decade/practice/controllers/rest/ChatController.kt
package com.decade.practice.controllers.rest

import com.decade.practice.core.ChatOperations
import com.decade.practice.database.repository.ChatRepository
import com.decade.practice.database.repository.UserRepository
import com.decade.practice.database.repository.get
import com.decade.practice.model.domain.ChatSnapshot
import com.decade.practice.model.domain.embeddable.ChatIdentifier
import com.decade.practice.utils.CacheUtils.defaultCacheControl
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/**
 * REST controller responsible for handling chat-related HTTP requests.
 * Provides endpoints for retrieving chat snapshots and chat lists.
 * Mapped to the "/chat" base URL path.
 */
@RestController
@RequestMapping("/chat")
class ChatController(
      private val userRepo: UserRepository,
      private val chatRepo: ChatRepository,
      private val chatOperations: ChatOperations,
) {

      /**
       * Retrieves a single chat snapshot based on the provided identifier.
       *
       * @param username The authenticated user's username extracted from the security principal
       * @param identifier The unique identifier for the chat, consisting of two user IDs
       * @param atVersion Optional version parameter to retrieve a specific chat snapshot version
       * @return ResponseEntity containing the chat snapshot with caching headers
       */
      @GetMapping("/snapshot")
      fun get(
            @AuthenticationPrincipal(expression = "name") username: String,
            @RequestParam @Validated identifier: ChatIdentifier,
            @RequestParam(required = false) atVersion: Int?
      ): ResponseEntity<ChatSnapshot> {

            // Configure cache control to improve performance
            val cacheControl = defaultCacheControl
            // Retrieve the authenticated user
            val me = userRepo.getByUsername(username)
            // Get or create the chat based on the identifier
            val chat = chatOperations.getOrCreateChat(identifier)
            // Generate a snapshot at the specified version or use the user's current sync version
            val snapshot = chatOperations.getSnapshot(chat, me, atVersion ?: me.syncContext.eventVersion)

            // Return the snapshot with appropriate cache headers
            return ResponseEntity.ok().cacheControl(cacheControl)
                  .varyBy("Cookie", "Authorization")
                  .body(snapshot)
      }

      /**
       * Retrieves a list of chat snapshots for the authenticated user.
       *
       * @param username The authenticated user's username extracted from the security principal
       * @param startAt Optional chat identifier to start the list from (for pagination)
       * @param atVersion The version of chat snapshots to retrieve
       * @return ResponseEntity containing a list of chat snapshots with caching headers
       */
      @GetMapping
      fun list(
            @AuthenticationPrincipal(expression = "name") username: String,
            @RequestParam(required = false) @Validated startAt: ChatIdentifier?,
            @RequestParam atVersion: Int,
      ): ResponseEntity<List<ChatSnapshot>> {
            // Get the starting chat if startAt is provided, otherwise use null
            val chat = if (startAt == null) null else chatRepo.get(startAt)
            // Retrieve the authenticated user
            val owner = userRepo.getByUsername(username)

            // Configure cache control to improve performance
            val cacheControl = defaultCacheControl
            // Get the list of chats for the user
            val chatList = chatOperations.listChat(owner, atVersion, chat)
            // Transform each chat into a snapshot at the specified version
            val snapshotList = chatList.map { chat ->
                  chatOperations.getSnapshot(chat, owner, atVersion)
            }

            // Return the list of snapshots with appropriate cache headers
            return ResponseEntity.ok().cacheControl(cacheControl)
                  .varyBy("Cookie", "Authorization")
                  .body(snapshotList)
      }
}